core/Enclave/enclave_hsm.cpp

```diff
@@ -202,6 +235,227 @@ sgx_status_t enclave_create_key(ehsm_keyblob_t *cmk, size_t cmk_size)
+                                         ehsm_padding_mode_t padding_mode,
+                                         ehsm_data_t *key_material, size_t key_material_size)
+{
+    sgx_status_t ret = SGX_ERROR_UNEXPECTED;
+
+    if (cmk == NULL ||
+        cmk_size != APPEND_SIZE_TO_KEYBLOB_T(cmk->keybloblen) ||
+        cmk->metadata.origin != EH_EXTERNAL_KEY ||
+        key_material_size != APPEND_SIZE_TO_DATA_T(key_material->datalen) ||
+        (cmk->metadata.keyusage != EH_KEYUSAGE_ENCRYPT_DECRYPT && cmk->metadata.keyusage != EH_KEYUSAGE_SIGN_VERIFY))
+    {
+        return SGX_ERROR_INVALID_PARAMETER;
+    }
+
+    ehsm_data_t *import_key = NULL;
+    int retval = 0;
+
+    int paddingMode = getPaddingMode(padding_mode);
+    if (paddingMode != RSA_PKCS1_PADDING && paddingMode != RSA_PKCS1_OAEP_PADDING)
+        return SGX_ERROR_INVALID_PARAMETER;
+
+    uint8_t *rsa_keypair = NULL;
+    BIO *bio = NULL;
+    RSA *rsa_prikey = NULL;
+
+    uint32_t keylength = ehsm_get_gcm_ciphertext_size((sgx_aes_gcm_data_ex_t *)cmk->keyblob);
+    switch (keylength)
+    {
+    case 0:
+    case 16:
+    case 24:
+    case 32:
+        ret = SGX_ERROR_INVALID_STATE;
+        goto out;
+    default:
+        break;
+    }
+
+    // load private key
+    rsa_keypair = (uint8_t *)malloc(cmk->keybloblen);
+    if (rsa_keypair == NULL)
+        goto out;
+
+    ret = ehsm_parse_keyblob(rsa_keypair,
+                             (sgx_aes_gcm_data_ex_t *)cmk->keyblob);
+    if (ret != SGX_SUCCESS)
+        goto out;
+
+    bio = BIO_new_mem_buf(rsa_keypair, -1); // use -1 to auto compute length
+    if (bio == NULL)
+    {
+        log_e("failed to load key pem\n");
+        ret = SGX_ERROR_UNEXPECTED;
+        goto out;
+    }
+
+    PEM_read_bio_RSAPrivateKey(bio, &rsa_prikey, NULL, NULL);
+    if (rsa_prikey == NULL)
+    {
+        log_e("failed to load private key\n");
+        ret = SGX_ERROR_UNEXPECTED;
+        goto out;
+    }
+
+    import_key = (ehsm_data_t *)malloc(APPEND_SIZE_TO_DATA_T(RSA_size(rsa_prikey)));
+    memset(import_key, 0, APPEND_SIZE_TO_DATA_T(RSA_size(rsa_prikey)));
+
+    retval = RSA_private_decrypt(key_material->datalen,
+                                 key_material->data,
+                                 import_key->data,
+                                 rsa_prikey,
+                                 paddingMode);
+    if (retval <= 0)
+    {
+        log_e("failed to make rsa decrypt\n");
+        ret = SGX_ERROR_UNEXPECTED;
+        goto out;
+    }
+    import_key->datalen = retval;
+
+    ret = check_aeskey_length(retval, cmk->metadata.keyspec);
+    if (ret != SGX_SUCCESS)
+        goto out;
+
+    memset_s(cmk->keyblob, cmk->keybloblen, 0, cmk->keybloblen);
+
+    ret = ehsm_create_keyblob(import_key->data,
+                              import_key->datalen,
+                              (sgx_aes_gcm_data_ex_t *)cmk->keyblob);
+
+    cmk->keybloblen = import_key->datalen + sizeof(sgx_aes_gcm_data_ex_t);
```

Comment:
Is it possible the memory allocated before is not sufficient to store? 
If we support 64B key, how about here ?